editor/src/components/editor/store/store-deep-equality-instances.ts

```diff
@@ -64,3 +136,508 @@ export function DerivedStateKeepDeepEquality(): KeepDeepEqualityCall<DerivedStat
+  if (oldPoint.x === newPoint.x && oldPoint.y === newPoint.y) {
+    return keepDeepEqualityResult(oldPoint, true)
+  } else {
+    return keepDeepEqualityResult(newPoint, false)
+  }
+}
+
+export function SidesKeepDeepEquality(
+  oldSides: Sides,
+  newSides: Sides,
+): KeepDeepEqualityResult<Sides> {
+  if (
+    oldSides.top === newSides.top &&
+    oldSides.left === newSides.left &&
+    oldSides.right === newSides.right &&
+    oldSides.bottom === newSides.bottom
+  ) {
+    return keepDeepEqualityResult(oldSides, true)
+  } else {
+    return keepDeepEqualityResult(newSides, false)
+  }
+}
+
+export function SpecialSizeMeasurementsKeepDeepEquality(): KeepDeepEqualityCall<
+  SpecialSizeMeasurements
+> {
+  return (oldSize, newSize) => {
+    const offsetResult = LocalPointKeepDeepEquality(oldSize.offset, newSize.offset)
+    const coordinateSystemBoundsResult = nullableDeepEquality(CanvasRectangleKeepDeepEquality)(
+      oldSize.coordinateSystemBounds,
+      newSize.coordinateSystemBounds,
+    )
+    const immediateParentBoundsResult = nullableDeepEquality(CanvasRectangleKeepDeepEquality)(
+      oldSize.immediateParentBounds,
+      newSize.immediateParentBounds,
+    )
+    const immediateParentProvidesLayoutResult =
+      oldSize.immediateParentProvidesLayout === newSize.immediateParentProvidesLayout
+    const usesParentBoundsResult = oldSize.usesParentBounds === newSize.usesParentBounds
+    const parentLayoutSystemResult = oldSize.parentLayoutSystem === newSize.parentLayoutSystem
+    const layoutSystemForChildrenResult =
+      oldSize.layoutSystemForChildren === newSize.layoutSystemForChildren
+    const providesBoundsForChildrenResult =
+      oldSize.providesBoundsForChildren === newSize.providesBoundsForChildren
+    const positionResult = oldSize.position === newSize.position
+    const marginResult = SidesKeepDeepEquality(oldSize.margin, newSize.margin)
+    const paddingResult = SidesKeepDeepEquality(oldSize.padding, newSize.padding)
+    const naturalWidthResult = oldSize.naturalWidth === newSize.naturalWidth
+    const naturalHeightResult = oldSize.naturalHeight === newSize.naturalHeight
+    const clientWidthResult = oldSize.clientWidth === newSize.clientWidth
+    const clientHeightResult = oldSize.clientHeight === newSize.clientHeight
+    const parentFlexDirectionResult = oldSize.parentFlexDirection === newSize.parentFlexDirection
+    const areEqual =
+      offsetResult.areEqual &&
+      coordinateSystemBoundsResult.areEqual &&
+      immediateParentBoundsResult.areEqual &&
+      immediateParentProvidesLayoutResult &&
+      usesParentBoundsResult &&
+      parentLayoutSystemResult &&
+      layoutSystemForChildrenResult &&
+      providesBoundsForChildrenResult &&
+      positionResult &&
+      marginResult.areEqual &&
+      paddingResult.areEqual &&
+      naturalWidthResult &&
+      naturalHeightResult &&
+      clientWidthResult &&
+      clientHeightResult &&
+      parentFlexDirectionResult
```

Comment:
`SpecialSizeMeasurements` is one of those objects which we change the shape pretty often (mostly to extend it, to be honest), it would be important to use some sort of type enforcement that makes sure we are not leaving this code behind. why is this not using the `combineNEqualityCalls` function?

Reply:
Because I didn't want to make a `combine16EqualityCalls` function and most of the values in here are primitives so they don't need any particularly special handling.

Reply:
I worry about friction making things bad here. how about writing a variadic combiner function? 

my problem is that if we add new parameters to the list of SpecialSizeMeasurements, not only the code will compile without letting us know that this needs to be updated, the chance that we catch a bug before the PR is merged is very low, because those new parameters need to be the only ones changing.

Reply:
The code will only compile if a default parameter is used in `specialSizeMeasurements`, which I think is a worse anti-pattern that catches us out surprisingly often.